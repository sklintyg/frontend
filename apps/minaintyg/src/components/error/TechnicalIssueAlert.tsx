import { ErrorIdentifier } from '@frontend/components'
import { IDSAlert } from '@frontend/ids-react-ts'
import { ReactNode } from 'react'
import { QueryError } from '../../utils/isQueryError'

export function TechnicalIssueAlert({
  children,
  headline = 'Tekniskt fel',
  error,
  additionalInfo,
}: {
  children: ReactNode
  headline?: string
  error: QueryError
  additionalInfo?: ReactNode
}) {
  return (
    <IDSAlert headline={headline} type="error">
      <div className="flex flex-col gap-4">
        {children}
        {error.id && (
          <p>
            <ErrorIdentifier id={error.id} centerText={false} />
          </p>
        )}
        {additionalInfo && additionalInfo}
      </div>
    </IDSAlert>
  )
}
