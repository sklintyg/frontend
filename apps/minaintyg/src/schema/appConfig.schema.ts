import { z } from 'zod'

export const appConfigLinkSchema = z.object({
  key: z.string(),
  url: z.string(),
  text: z.string(),
  target: z.string(),
})

export const appConfigSchema = z.object({
  useMinifiedJavascript: z.boolean(),
  version: z.string(),
  buildNumber: z.string(),
  elva77MainUrl: z.string(),
  elva77LoginUrl: z.string(),
  applicationLogoutUrl: z.string(),
  miUserSurveyUrl: z.string(),
  miUserSurveyVersion: z.string(),
  miUserSurveyDateFrom: z.string(),
  miUserSurveyDateTo: z.string(),
  knownRecipients: z.array(z.object({ id: z.string(), name: z.string(), trusted: z.boolean() })),
  links: z.record(appConfigLinkSchema),
  banners: z.array(z.unknown()),
})

export type AppConfig = z.infer<typeof appConfigSchema>
export type AppConfigLink = z.infer<typeof appConfigLinkSchema>
