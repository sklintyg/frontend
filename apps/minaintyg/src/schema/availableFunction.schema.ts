import { z } from 'zod'

export const AvailableFunctionType = z.enum(['SEND_CERTIFICATE'])
export const AvailableFunctionInformationType = z.enum([''])

export const availableFunctionInformationSchema = z.object({
  type: AvailableFunctionInformationType,
  id: z.string(),
})

export const availableFunctionSchema = z.object({
  body: z.string().nullable(),
  description: z.string().nullable(),
  name: z.string().nullable(),
  title: z.string().nullable(),
  type: AvailableFunctionType,
  information: z.array(availableFunctionInformationSchema),
})

export type AvailableFunction = z.infer<typeof availableFunctionSchema>
