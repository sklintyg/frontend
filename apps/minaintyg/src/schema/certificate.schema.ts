import { z } from 'zod'

export const CertificateStatusEnum = z.enum(['NEW', 'REPLACED', 'SENT', 'NOT_SENT'])

export const certificateEventSchema = z.object({
  timestamp: z.string().datetime(),
  certificateId: z.string(),
  description: z.string(),
})

export const certificateSummarySchema = z.object({ label: z.string(), value: z.string() }).optional()

export const certificateTypeSchema = z.object({
  id: z.string(),
  version: z.string(),
  description: z.string().optional(),
})

export const certificateIssuerSchema = z.object({
  name: z.string(),
  phoneNumber: z.string(),
})

export const certificateUnitSchema = z.object({
  id: z.string(),
  name: z.string(),
  address: z.string(),
})

export const certificateMetadataSchema = z.object({
  issuer: certificateIssuerSchema,
  unit: certificateUnitSchema,
  name: z.string(),
  events: z.array(certificateEventSchema),
  statuses: z.array(CertificateStatusEnum),
  id: z.string(),
  issued: z.string().datetime(),
  type: certificateTypeSchema,
  summary: certificateSummarySchema,
})

export const certificateContentSchema = z.object({ heading: z.string(), body: z.string() })

export const certificateSchema = z.object({
  metadata: certificateMetadataSchema,
  content: z.array(certificateContentSchema),
})

export type CertificateContent = z.infer<typeof certificateContentSchema>
export type CertificateEvent = z.infer<typeof certificateEventSchema>
export type CertificateIssuer = z.infer<typeof certificateIssuerSchema>
export type CertificateMetadata = z.infer<typeof certificateMetadataSchema>
export type CertificateStatus = z.infer<typeof CertificateStatusEnum>
export type CertificateSummary = z.infer<typeof certificateSummarySchema>
export type CertificateUnit = z.infer<typeof certificateUnitSchema>
export type Certificate = z.infer<typeof certificateSchema>
