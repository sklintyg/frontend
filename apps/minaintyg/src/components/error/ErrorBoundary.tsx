import { randomUUID } from '@frontend/utils'
import { useMemo } from 'react'
import { ErrorPageHero } from './ErrorPageHero'

export function ErrorBoundary() {
  const id = useMemo(randomUUID, [])
  // TODO: useEffect hook that logs error back to the backend.
  return <ErrorPageHero id={id} type="unknown" />
}
