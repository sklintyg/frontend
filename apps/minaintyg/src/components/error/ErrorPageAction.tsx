import { getNavigationItem, getNavigationItemUrl } from '@frontend/components/1177'
import { IDSIconChevron, IDSIconExternal, IDSLink } from '@frontend/ids-react-ts'
import { Link } from 'react-router-dom'
import type { ErrorPageActionTypeEnum } from '../../schema/error.schema';
import { ErrorPageActionType } from '../../schema/error.schema'

export function ErrorPageAction({ type }: { type: ErrorPageActionTypeEnum }) {
  const startLinkItem = getNavigationItem('Start')

  if (type === ErrorPageActionType.enum.start) {
    return (
      startLinkItem && (
        <IDSLink key="start">
          <IDSIconChevron />
          <Link to={getNavigationItemUrl(startLinkItem, import.meta.env.MODE)}>Till startsidan</Link>
        </IDSLink>
      )
    )
  }

  if (type === ErrorPageActionType.enum.login) {
    return (
      <IDSLink key="login">
        <IDSIconChevron />
        <a href="/">Till inloggning</a>
      </IDSLink>
    )
  }

  if (type === ErrorPageActionType.enum[1177]) {
    return (
      <IDSLink key="inera">
        <IDSIconChevron />
        <Link to="https://www.1177.se">Till 1177</Link>
        <IDSIconExternal slot="append-icon" />
      </IDSLink>
    )
  }

  if (type === ErrorPageActionType.enum.about) {
    return (
      <IDSLink key="inera">
        <IDSIconChevron />
        <Link to="https://www.1177.se/om-1177/nar-du-loggar-in-pa-1177.se/det-har-kan-du-gora-nar-du-loggat-in/hantera-intyg/hantera-dina-lakarintyg-och-lakarutlatanden/">
          Om 1177 intyg
        </Link>
        <IDSIconExternal slot="append-icon" />
      </IDSLink>
    )
  }
}
