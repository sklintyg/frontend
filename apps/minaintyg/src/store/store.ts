import { configureStore } from '@reduxjs/toolkit'
import { api } from './api'
import { errorMiddleware } from './middleware/error.middleware'
import { reducer } from './reducer'
import { idpMiddleware } from './slice/idpMiddleware'

export const store = configureStore({
  reducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([api.middleware, idpMiddleware, errorMiddleware]),
})

export type AppDispatch = typeof store.dispatch
