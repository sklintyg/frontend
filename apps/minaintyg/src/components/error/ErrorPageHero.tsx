import { ErrorIdentifier } from '@frontend/components'
import { PageHero, PageHeroActions } from '@frontend/components/1177'
import { ErrorPageActionType, ErrorType, ErrorTypeEnum } from '../../schema/error.schema'
import { ErrorPageAction } from './ErrorPageAction'
import { SupportLink } from './SupportLink/SupportLink'

const errorInfo = {
  [ErrorType.enum.unavailable]: {
    heading: 'Tjänsten är inte tillgänglig just nu',
    description: 'Välkommen tillbaka vid senare tillfälle.',
    actions: [],
  },
  [ErrorType.enum['login-failed']]: {
    heading: 'Inloggning misslyckades',
    description: 'På grund av ett tekniskt fel går det inte att logga in just nu. Försök igen senare.',
    actions: [ErrorPageActionType.enum.start],
  },
  [ErrorType.enum['logged-out']]: {
    heading: 'Du är utloggad',
    description:
      'Du har antingen valt att logga ut eller blivit utloggad på grund av inaktivitet. Du kan nu stänga fliken eller logga in på nytt.',
    actions: [ErrorPageActionType.enum.login, ErrorPageActionType.enum[1177]],
  },
  [ErrorType.enum['not-found']]: {
    heading: 'Sidan hittades inte',
    description: 'Sidan du söker har fått en ny adress eller är borttagen.',
    actions: [ErrorPageActionType.enum.start],
  },
  [ErrorType.enum.underage]: {
    heading: 'Du som är under 16 har inte tillgång till 1177 intyg',
    description: 'Vänd dig till din mottagning om du har frågor eller behöver en kopia av ditt intyg.',
    actions: [ErrorPageActionType.enum.start, ErrorPageActionType.enum.about],
  },
  [ErrorType.enum.unknown]: {
    heading: 'Någonting gick fel',
    description: 'På grund av ett tekniskt fel går det inte att visa Intyg just nu. Försök igen senare.',
    actions: [ErrorPageActionType.enum.start],
  },
} as const

function getErrorHeading(type?: ErrorTypeEnum) {
  return (type !== undefined && errorInfo[type]?.heading) || errorInfo.unknown.heading
}

function getErrorDescription(type?: ErrorTypeEnum) {
  return (type !== undefined && errorInfo[type]?.description) || errorInfo.unknown.description
}

function getActions(type?: ErrorTypeEnum) {
  return (type !== undefined && errorInfo[type]?.actions) || errorInfo.unknown.actions
}

export function ErrorPageHero({ id, type }: { id?: string; type?: ErrorTypeEnum }) {
  return (
    <PageHero heading={getErrorHeading(type)} type={type === ErrorType.enum['logged-out'] ? 'success' : 'error'}>
      <p className="ids-preamble">{getErrorDescription(type)}</p>
      <PageHeroActions>
        {getActions(type).map((t) => (
          <ErrorPageAction type={t} key={t} />
        ))}
      </PageHeroActions>
      {id && type !== ErrorType.enum['logged-out'] && (
        <div className="border-y border-stone-clear py-5">
          Om problemet kvarstår, spara nedan id och kontakta <SupportLink />
          <br />
          <ErrorIdentifier id={id} showTitle={false} />
        </div>
      )}
    </PageHero>
  )
}
