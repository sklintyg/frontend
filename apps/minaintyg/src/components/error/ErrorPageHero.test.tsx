import { faker } from '@frontend/fake'
import { render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'
import { ErrorType } from '../../schema/error.schema'
import { store } from '../../store/store'
import { ErrorPageHero } from './ErrorPageHero'

it('Should render identifier', () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <ErrorPageHero id="037b5ea0-ca66-49dd-abfa-a498dc0b1027" />
      </MemoryRouter>
    </Provider>
  )
  expect(screen.getByText('Om problemet kvarstÃ¥r, spara nedan id och kontakta')).toBeInTheDocument()
  expect(screen.getByRole('link', { name: 'support' })).toBeInTheDocument()
  expect(screen.getByText('037b5ea0-ca66-49dd-abfa-a498dc0b1027')).toBeInTheDocument()
})

it.each(ErrorType.options)('Should render %s type error', (type) => {
  faker.seed(1234)
  const { container } = render(
    <Provider store={store}>
      <MemoryRouter>
        <ErrorPageHero type={type} />
      </MemoryRouter>
    </Provider>
  )
  expect(container).toMatchSnapshot()
})
