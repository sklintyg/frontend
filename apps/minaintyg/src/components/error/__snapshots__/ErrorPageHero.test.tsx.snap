// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render logged-out type error 1`] = `
<div>
  <div
    class="ids-content mx-auto flex max-w-[620px] flex-col gap-10 pt-12 text-center md:pt-24"
  >
    <div
      class="mx-auto"
      style="font-size: 72px;"
    >
      <span
        class="ids-icon-success "
      />
    </div>
    <div>
      <h1
        class="ids-heading-xxl !normal-case text-center"
      >
        Du är utloggad
      </h1>
      <p
        class="ids-preamble"
      >
        Du har antingen valt att logga ut eller blivit utloggad på grund av inaktivitet. Du kan nu stänga fliken eller logga in på nytt.
      </p>
      <div
        class="mb-5 text-left md:text-center"
      >
        <div
          class="mb-5 mt-8 w-[43px] border-b md:hidden"
        />
        <div
          class="mx-auto inline-flex flex-col gap-2.5 md:flex-row"
        >
          <a
            class="ids-link"
            href="/"
            rel="noreferrer"
            target="_blank"
          >
            <span
              class="ids-icon-arrow-right-small  ids-icon--text-start"
            />
            Till inloggning
            <span
              class="ids-icon-external-link-small  ids-icon--text-end"
            />
          </a>
          <a
            class="ids-link"
            href="https://www.1177.se"
            rel="noreferrer"
            target="_blank"
          >
            <span
              class="ids-icon-arrow-right-small  ids-icon--text-start"
            />
            Till 1177
            <span
              class="ids-icon-external-link-small  ids-icon--text-end"
            />
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render login-failed type error 1`] = `
<div>
  <div
    class="ids-content mx-auto flex max-w-[620px] flex-col gap-10 pt-12 text-center md:pt-24"
  >
    <div
      class="mx-auto"
      style="font-size: 72px;"
    >
      <span
        class="ids-icon-attention "
      />
    </div>
    <div>
      <h1
        class="ids-heading-xxl !normal-case text-center"
      >
        Inloggning misslyckades
      </h1>
      <p
        class="ids-preamble"
      >
        På grund av ett tekniskt fel går det inte att logga in just nu. Försök igen senare.
      </p>
      <div
        class="mb-5 text-left md:text-center"
      >
        <div
          class="mb-5 mt-8 w-[43px] border-b md:hidden"
        />
        <div
          class="mx-auto inline-flex flex-col gap-2.5 md:flex-row"
        >
          <a
            class="ids-link"
            href="https://e-tjanster.1177.se/mvk/"
          >
            <span
              class="ids-icon-arrow-right-small  ids-icon--text-start"
            />
            Till startsidan
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render not-found type error 1`] = `
<div>
  <div
    class="ids-content mx-auto flex max-w-[620px] flex-col gap-10 pt-12 text-center md:pt-24"
  >
    <div
      class="mx-auto"
      style="font-size: 72px;"
    >
      <span
        class="ids-icon-attention "
      />
    </div>
    <div>
      <h1
        class="ids-heading-xxl !normal-case text-center"
      >
        Sidan hittades inte
      </h1>
      <p
        class="ids-preamble"
      >
        Sidan du söker har fått en ny adress eller är borttagen.
      </p>
      <div
        class="mb-5 text-left md:text-center"
      >
        <div
          class="mb-5 mt-8 w-[43px] border-b md:hidden"
        />
        <div
          class="mx-auto inline-flex flex-col gap-2.5 md:flex-row"
        >
          <a
            class="ids-link"
            href="https://e-tjanster.1177.se/mvk/"
          >
            <span
              class="ids-icon-arrow-right-small  ids-icon--text-start"
            />
            Till startsidan
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render unavailable type error 1`] = `
<div>
  <div
    class="ids-content mx-auto flex max-w-[620px] flex-col gap-10 pt-12 text-center md:pt-24"
  >
    <div
      class="mx-auto"
      style="font-size: 72px;"
    >
      <span
        class="ids-icon-attention "
      />
    </div>
    <div>
      <h1
        class="ids-heading-xxl !normal-case text-center"
      >
        Tjänsten är inte tillgänglig just nu
      </h1>
      <p
        class="ids-preamble"
      >
        Välkommen tillbaka vid senare tillfälle.
      </p>
    </div>
  </div>
</div>
`;

exports[`Should render underage type error 1`] = `
<div>
  <div
    class="ids-content mx-auto flex max-w-[620px] flex-col gap-10 pt-12 text-center md:pt-24"
  >
    <div
      class="mx-auto"
      style="font-size: 72px;"
    >
      <span
        class="ids-icon-attention "
      />
    </div>
    <div>
      <h1
        class="ids-heading-xxl !normal-case text-center"
      >
        Du som är under 16 har inte tillgång till 1177 intyg
      </h1>
      <p
        class="ids-preamble"
      >
        Vänd dig till din mottagning om du har frågor eller behöver en kopia av ditt intyg.
      </p>
      <div
        class="mb-5 text-left md:text-center"
      >
        <div
          class="mb-5 mt-8 w-[43px] border-b md:hidden"
        />
        <div
          class="mx-auto inline-flex flex-col gap-2.5 md:flex-row"
        >
          <a
            class="ids-link"
            href="https://e-tjanster.1177.se/mvk/"
          >
            <span
              class="ids-icon-arrow-right-small  ids-icon--text-start"
            />
            Till startsidan
          </a>
          <a
            class="ids-link"
            href="https://www.1177.se/om-1177/nar-du-loggar-in-pa-1177.se/det-har-kan-du-gora-nar-du-loggat-in/hantera-intyg/hantera-dina-lakarintyg-och-lakarutlatanden/"
            rel="noreferrer"
            target="_blank"
          >
            <span
              class="ids-icon-arrow-right-small  ids-icon--text-start"
            />
            Om 1177 intyg
            <span
              class="ids-icon-external-link-small  ids-icon--text-end"
            />
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render unknown type error 1`] = `
<div>
  <div
    class="ids-content mx-auto flex max-w-[620px] flex-col gap-10 pt-12 text-center md:pt-24"
  >
    <div
      class="mx-auto"
      style="font-size: 72px;"
    >
      <span
        class="ids-icon-attention "
      />
    </div>
    <div>
      <h1
        class="ids-heading-xxl !normal-case text-center"
      >
        Någonting gick fel
      </h1>
      <p
        class="ids-preamble"
      >
        På grund av ett tekniskt fel går det inte att visa Intyg just nu. Försök igen senare.
      </p>
      <div
        class="mb-5 text-left md:text-center"
      >
        <div
          class="mb-5 mt-8 w-[43px] border-b md:hidden"
        />
        <div
          class="mx-auto inline-flex flex-col gap-2.5 md:flex-row"
        >
          <a
            class="ids-link"
            href="https://e-tjanster.1177.se/mvk/"
          >
            <span
              class="ids-icon-arrow-right-small  ids-icon--text-start"
            />
            Till startsidan
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
